sdk/src/test/java/com/akkaserverless/javasdk/replicatedentity/CartEntity.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2021 Lightbend Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.akkaserverless.javasdk.replicatedentity;
+
+import com.example.replicatedentity.shoppingcart.ShoppingCartApi;
+import com.example.replicatedentity.shoppingcart.domain.ShoppingCartDomain;
+import com.google.protobuf.Empty;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class CartEntity extends AbstractCartEntity {
+  @SuppressWarnings("unused")
+  private final String entityId;
+
+  public CartEntity(ReplicatedEntityContext context) {
+    this.entityId = context.entityId();
+  }
+
+  @Override
+  public ReplicatedRegisterMap<String, ShoppingCartDomain.LineItem> emptyData(
+      ReplicatedDataFactory factory) {
+    return factory.newReplicatedRegisterMap();
+  }
+
+  @Override
+  public Effect<Empty> addItem(
+      ReplicatedRegisterMap<String, ShoppingCartDomain.LineItem> currentData,
+      ShoppingCartApi.AddLineItem addLineItem) {
+    if (addLineItem.getQuantity() <= 0) {
+      return effects().error("Cannot add negative quantity to item " + addLineItem.getProductId());
+    }
+
+    currentData.setValue(addLineItem.getProductId(), updateItem(addLineItem, currentData));
```

Comment:
I didn't follow the API discussions on replicated, so maybe I missed something, but I was a bit surprised that it is based on mutating the structure and not returning it as a part of the effect like for the value/es entities.

Reply:
Right, I haven't changed that part of the API in this PR. Just removing the annotation/reflection/context-based approach for the entity as a whole. No discussions yet, as far as I know, around what an immutable or effect-based API for the CRDTs themselves should be.

We could make the replicated data immutable, where the new structure is passed through to an update effect; or we could have commands for effects (specific to the replicated data type), like `Increment` or `SetValue`.

Is there a particular approach that you're thinking of already?

Reply:
I think a "managed" immutable data structure would better match the value/es entities and be less confusing vs a mutable object passed to the method to mutate to achieve the change. Even with a mutable structure I think having an effect saying "update this" could make sense for clarity.

Reply:
Yeah, the immutable data structure would be a step towards something more similar.

Another approach would be to rework things completely, so rather than having a replicated entity that contains replicated data object, there are specialised replicated entities. It's now `ReplicatedEntity<ReplicatedCounter>`, while it could as well be `ReplicatedCounterEntity`. So currently, an immutable-style API would be along the lines of:

```java
class MyCounter extends AbstractCounter { // extends ReplicatedEntity<ReplicatedCounter>
  @Override public ReplicatedCounter emptyData(ReplicatedDataFactory factory) {
    return factory.newCounter();
  }

  @Override public Effect<Response> someMethod(ReplicatedCounter counter, Request request) {
    return effects()
        .update(counter.increment(request.getValue()))
        .thenReply(Response.newBuilder...);
  }
}
```

While something that's a specialised entity for each replicated data type could instead be closer to event sourced or value entities:

```java
class MyCounter extends AbstractCounter { // extends ReplicatedCounterEntity
  // no need for empty data — counter empty state is 0 already
  // register can have an empty state
  
  // pass in the current replicated data value — Long, Set, Map, etc
  @Override public Effect<Response> someMethod(long currentValue, Request request) {
    return effects()
        .increment(request.getValue()) // specific effects for each replicated data type
        .thenReply(Response.newBuilder...);
  }
}
```

Reply:
I think that looks very nice with the specialized entity for the data type.

Reply:
That looks good, but how much do we have to duplicate/specialize in the different parts of codegen and abstract classes?

The immutable + effects.update isn't too bad either.

Reply:
> how much do we have to duplicate/specialize in the different parts of codegen and abstract classes

I think this will be similar for any of the approaches, where codegen will need to be specific to the replicated data type. It would just be the difference between using the right type parameter or the right base class.

I'll move the API discussions for the replicated data level over to #83.